### PR TITLE
Split workflow in several hooks

### LINKED ISSUE #14: Split codecoverage bot into multiple hooks
Right now we have one hook which is then doing different things according to the input. We should instead have separate hooks.

### PR DESCRIPTION
**Pull request by La0**
Fixes #14 

So we can work on https://github.com/mozilla/code-coverage/issues/150

### CODE REVIEW

**Comment by marco-c** [comment]
Could you split it in actual separate Taskcluster hooks? One for cron, one triggered by pulselistener. Then, you could add an argument to cli.py to choose the workflow, to make it less brittle than relying on the values of other args (https://github.com/mozilla/code-coverage/pull/165/files#diff-6310014f6dc184cfe909a867e61bfcb8R64-R77).

Nit: I'd rename "mc.py" and "try_repo.py" to make it clear what they do (e.g. if we support other repos in the future, "mc.py" would not make sense anymore). "mc.py" is the task that runs at the end of a full test run, "try_repo.py" is the one that runs on-demand when developers push to try. I'm not sure what a good naming would be, maybe tests.py/all_tests.py/full_tests.py/push_tests.py/push.py/merge.py/merge_tests.py and on_demand.py?

File: bot/code_coverage_bot/hooks/try_repo.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+        self.retrieve_source_and_artifacts()
+
+        reports = self.build_reports(only=[("all", "all")])
+        full_path = reports.get(("all", "all"))
+        assert full_path is not None, "Missing full report (all:all)"
+        report = json.load(open(full_path))
+
+        logger.info("Upload changeset coverage data to Phabricator")
+        phabricatorUploader.upload(report, changesets)
```
Comment (marco-c) [inline bot/code_coverage_bot/hooks/try_repo.py]: A lot of this code is shared between mc and try_repo, maybe you could move it to a function (either in the base class or refactoring PhabricatorUploader

File: bot/code_coverage_bot/codecov.py [hunk 2]
```diff
@@ -1,303 +0,0 @@
-# -*- coding: utf-8 -*-
```
Comment (marco-c) [inline bot/code_coverage_bot/codecov.py]: It's a pity that git doesn't realize this file was partially moved :(

**Comment by La0** [comment]
I went further than your comment, and used only 2 hooks:
1. `code_review_bot.hooks.repo` for _all_ repositories, using a static configuration to change behaviour according to the repository in use
2. `code_review_bot.hooks.cron` for a scheduled task

I prefered to use 2 different endpoints, as some CLI options are not used by the cron worfklow.
The CLI & TC setup code is shared by both endpoints though.

In the end, i think the code is a bit cleaner, as we now only have 2 different worfklows (one being parametrizable), and we can support other repos a bit more easily.

File: bot/code_coverage_bot/hooks/repo.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+from code_coverage_bot.cli import setup_cli
+from code_coverage_bot.hooks.base import Hook
+from code_coverage_bot.notifier import notify_email
+from code_coverage_bot.phabricator import PhabricatorUploader
+from code_coverage_bot.phabricator import parse_revision_id
+
+logger = structlog.get_logger(__name__)
+
+
+REPOSITORIES = {
```
[the remark this answers is not in the document]
Comment (La0) [reply bot/code_coverage_bot/hooks/repo.py]: Agreed, and done :smile:

File: bot/code_coverage_bot/hooks/repo.py [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+
+    def run(self):
+        # Check the covdir report does not already exists
+        if self.config["gcp_upload"] and uploader.gcp_covdir_exists(
+            self.branch, self.revision, "all", "all"
+        ):
+            logger.warn("Full covdir report already on GCP")
+            return
+
+        self.retrieve_source_and_artifacts()
```
[the remark this answers is not in the document]
Comment (La0) [reply bot/code_coverage_bot/hooks/repo.py]: I re-applied the original workflow

**Pull request merged**